fix: validate last cosmos step on broadcast

NeOMakinG opened the pull request:
## Description
- The last step wasn't success on broadcast
<!-- Please describe your changes -->

## Notice

<!-- Before submitting a pull request, please make sure you have answered the following: -->

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [x] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)
closes #1553
<!-----------------------------------------------------------------------------
If applicable, please link to the github issue and put `closes #XXXX` in your comment to auto-close the issue that your PR fixes.
------------------------------------------------------------------------------>

## Risk
- No real risk
<!-----------------------------------------------------------------------------
Outline the scope of your changes and the risk associated with them. You must use your discretion as an engineer to determine the potential impact of your changes.

E.g. an upgrade to `hdwallet` or core state management would be considered higher risk, and might require a full regression test. UI or isolated view changes, or something behind a feature flag may have near zero risk. Small bug fixes might require testing isolated to the specific fix.
------------------------------------------------------------------------------>

## Testing
See video
<!-----------------------------------------------------------------------------
If your testing steps are technical, outline steps for an engineer to verify.

If your testing steps are functional, please provide a full guide with any features requiring special attention for operations to test your branch in the preview environment.
------------------------------------------------------------------------------>

## Screenshots (if applicable)
https://www.awesomescreenshot.com/video/8536924?key=98a0c129f859cbf8e2ea50fb0885dbf2

gomesalexandre commented on src/plugins/cosmos/components/modals/Staking/Staking.tsx:
```diff
@@ -39,9 +39,25 @@ import { UnstakeBroadcast } from './views/UnstakeBroadcast'
   const location = useLocation()
   const translate = useTranslate()
 
+  const isBroadcast = matchPath(location.pathname, {
```
Having some variable specifically for broadcast sounds a bit flaky, I'd prefer a way to systematically have the "step passed with green check" logic.

gomesalexandre commented on src/components/RouteSteps/RouteSteps.tsx:
```diff
@@ -21,17 +21,21 @@ type RouteStepsProps = {
   assetSymbol?: string
   action?: StakingAction
   location: Location
+  validateAllSteps?: boolean
```
The `validateXYZ` terminology is confusing. This isn't about validating anything, but rather about a step in the breadcrumb being done (in this case, the broadcast being successful, but I'd rather it being better abstracted than just being about the broadcast success)

gomesalexandre commented on src/plugins/cosmos/components/modals/Staking/Staking.tsx:
```diff
@@ -39,9 +39,25 @@ import { UnstakeBroadcast } from './views/UnstakeBroadcast'
+    }
+  }, [isBroadcast])
+
+  const onValidatedBroadcast = (value: boolean) => {
```
Ditto `validated` terminology, that's more like `onBroadcastSuccesful` (it is fired after a succesful broadcast Tx in `<ClaimBroadcast />`

gomesalexandre replied:
Why does this take a `value` arg if it's always called with `true`?

gomesalexandre commented on src/plugins/cosmos/components/modals/Staking/Staking.tsx:
```diff
@@ -41,32 +43,65 @@ import { UnstakeConfirm } from './views/UnstakeConfirm'
-    path: [ClaimPath.Confirm, ClaimPath.Broadcast],
-    exact: true,
-  })
+    setIsClaim(
+      matchPath(location.pathname, {
+        path: [ClaimPath.Confirm, ClaimPath.Broadcast],
+        exact: true,
+      }),
+    )
 
-  const isStake = matchPath(location.pathname, {
-    path: [StakeRoutes.Stake, StakingPath.Confirm, StakingPath.Broadcast],
-    exact: true,
-  })
+    setIsStake(
+      matchPath(location.pathname, {
+        path: [StakeRoutes.Stake, StakingPath.Confirm, StakingPath.Broadcast],
+        exact: true,
+      }),
+    )
 
-  const isUnstake = matchPath(location.pathname, {
-    path: [StakeRoutes.Unstake, UnstakingPath.Confirm, UnstakingPath.Broadcast],
-    exact: true,
-  })
+    setIsUnstake(
+      matchPath(location.pathname, {
+        path: [StakeRoutes.Unstake, UnstakingPath.Confirm, UnstakingPath.Broadcast],
+        exact: true,
+      }),
+    )
+  }, [location.pathname])
```
That's making things even more verbose than they were already.
Is there another way we could write this so we don't have to call all these state setters on this `useEffect`?
What if we use `<Route />`s  instead of `matchPatch`s, would it make things cleaner?

NeOMakinG replied:
We could by using `isUnstake?.isExact` as `dependencies` of the `setSteps` useEffect instead of the object directly, wdyt?

gomesalexandre replied:
Since we're going to eventually refactor/unify the routing of Foxy/DeFi/Cosmos, let's stamp this PR and resolve this.

gomesalexandre reviewed (commented):
```
I added an onStepCompleted method in order to complete the current step if needed (currently used in broadcast views)
```

That's still not adding *systematic* step success states. Now broadcast steps will have an `onStepCompleted` that denotes the `<step>/broadcast` substep being completed, but what about the steps before? We now effectively two bits of logic doing the same thing, in a different way, depending on if the step is one of the 3 broadcast substeps or any of the others. 

Also added a few more comments as less aggressive comments vs. changed planned since this is currently a draft, please take a look and let me know if there's anything unclear.

NeOMakinG commented:
> ```
> I added an onStepCompleted method in order to complete the current step if needed (currently used in broadcast views)
> ```
> 
> That's still not adding _systematic_ step success states. Now broadcast steps will have an `onStepCompleted` that denotes the `<step>/broadcast` substep being completed, but what about the steps before? We now effectively two bits of logic doing the same thing, in a different way, depending on if the step is one of the 3 broadcast substeps or any of the others.
> 
> Also added a few more comments as less aggressive comments vs. changed planned since this is currently a draft, please take a look and let me know if there's anything unclear.

Every others steps than broadcast including yearn and foxy are managed using the `location` object

I really don't get why we would want to add some logic to Cosmos previous steps if they are naturally working using `location`

If you want me to add the handler to any steps of the cosmos modal, just tell me, I'll do it but it feels useless to me

gomesalexandre commented on src/components/VerticalStepper/VerticalStepper.tsx:
```diff
@@ -24,7 +27,7 @@ export const VerticalStepper = ({ state, activeStep, steps }: VerticalStepperPro
       >
         {steps.map(step => (
-          <Step key={step.label} label={step.label} />
+          <Step key={step.label} label={step.label} isCompletedStep={step.isCompleted} />
```
Nitpick: Could just be `isCompleted`, the component is called `<Step />` so we're obviously talking about a step.

gomesalexandre reviewed (approved):
Looking much better after the naming and logic changes now!
Note: I haven't ran it locally yet, will do so in a bit as a second pass for this PR.

gomesalexandre commented:
@NeOMakinG As we've discussed, perfectly happy with this implementation for now, but as @elmutt suggested, it would be great to react on the actual Tx coming back from the broadcast vs. waiting on the broadcast promise.
This way, we would not only have stronger core, not shoehorned logic here, but I reckon this would also give us a lot more reliability on the send Txs for all chains.

What do you think? Could you add some TODOs here to make sure we remove this logic later on?

NeOMakinG commented:
> @NeOMakinG As we've discussed, perfectly happy with this implementation for now, but as @elmutt suggested, it would be great to react on the actual Tx coming back from the broadcast vs. waiting on the broadcast promise. This way, we would not only have stronger core, not shoehorned logic here, but I reckon this would also give us a lot more reliability on the send Txs for all chains.
> 
> What do you think? Could you add some TODOs here to make sure we remove this logic later on?

Yeah alright, no need to add todos these, you already did it while implementing cosmos in [both files](https://github.com/shapeshift/web/pull/1562/files#diff-50b4ef5311fcc7186f001e3257863906c0de06a74a3575a06d8555999bc14d18L72-L73)

gomesalexandre reviewed (approved):
Tested again, still working as intended.
Wasn't able to test unstake again since I'm on my max. undelegations on all validators currently, but looking at the diff it is exactly the same logic as stake/claim so I'll give you this: 🍬

gomesalexandre commented:
@NeOMakinG will merge this one after release
